src/main/kotlin/com/demonwav/mcdev/util/McBytecodeUtil.kt

```diff
@@ -0,0 +1,112 @@
+/*
+ * Minecraft Dev for IntelliJ
+ *
+ * https://minecraftdev.org
+ *
+ * Copyright (c) 2017 minecraft-dev
+ *
+ * MIT License
+ */
+
+@file:JvmName("McBytecodeUtil")
+package com.demonwav.mcdev.util
+
+import com.intellij.psi.PsiArrayType
+import com.intellij.psi.PsiClass
+import com.intellij.psi.PsiClassType
+import com.intellij.psi.PsiField
+import com.intellij.psi.PsiMethod
+import com.intellij.psi.PsiPrimitiveType
+import com.intellij.psi.PsiType
+import com.intellij.psi.util.TypeConversionUtil
+
+private const val INTERNAL_CONSTRUCTOR_NAME = "<init>"
+
+// Type
+
+internal val PsiPrimitiveType.internalName: Char
+    get() = when (this) {
+        PsiType.BYTE -> 'B'
+        PsiType.CHAR -> 'C'
+        PsiType.DOUBLE -> 'D'
+        PsiType.FLOAT -> 'F'
+        PsiType.INT -> 'I'
+        PsiType.LONG -> 'J'
+        PsiType.SHORT -> 'S'
+        PsiType.BOOLEAN -> 'Z'
+        PsiType.VOID -> 'V'
+        else -> throw IllegalArgumentException("Unsupported primitive type: $this")
+    }
+
+private fun PsiClassType.erasure() = TypeConversionUtil.erasure(this) as PsiClassType
+
+internal val PsiClassType.internalName
+    get() = erasure().resolve()!!.internalName
```

Comment:
The issue with your assumption is uncaught exceptions in IntelliJ don't fail silently. A bug notification pops up, with the stacktrace, and information on where the exception originated (this plugin).

> For example, directly below this property we have the appendInternalName method which takes a StringBuilder, appends the internal name and returns the builder again. I cannot return null for this method, so we would have one method which returns null and one method which throws an exception.

If it returns null then just don't append to the builder and return it? Either that, or throw a _specific_ exception so the actual failure can be handled, don't just throw a NPE (or cause an NPE to be thrown).

> how would we return null for the descriptor if one of the classes cannot be resolved?

In this case, like the above, a return value of `null` is either the fail case (which would then allow the _caller_ to handle the issue, not the utility method), or a specific exception should be thrown, rather than an ambiguous NPE. Either way, it should be in the contract of the method, and the caller should deal with it, but the caller should _never_ be catching a NPE because the utility method doesn't provide any other way to handle that error condition.